nemo/collections/nlp/models/language_modeling/megatron/falcon/falcon_decoder_layer.py

```diff
@@ -0,0 +1,284 @@
+from megatron.core.utils import make_viewless_tensor
+
+""" We use the following notation throughout this file:
+     h: hidden size
+     n: number of attention heads
+     p: number of model parallel partitions
+     np: n/p
+     hp: h/p
+     hn: h/n
+     b: batch size
+     s: sequence length
+     l: number of layers
+    Transformer takes input of size [s, b, h] and returns a
+    tensor of the same size. We use the following arguments:
+        hyperparameters: transformer hyperparameters
+"""
+
+
+@dataclass
+class FalconTransformerLayerSubmodules(LayerSubmodules):
+    input_layernorm: Union[ModuleSpec, type] = IdentityOp
+    self_attention: Union[ModuleSpec, type] = IdentityOp
+    self_attn_bda: Union[ModuleSpec, type] = IdentityFuncOp
+
+    post_self_attn_layernorm: Union[ModuleSpec, type] = IdentityOp
+
+    pre_mlp_layernorm: Union[ModuleSpec, type] = IdentityOp
+    mlp: Union[ModuleSpec, type] = IdentityOp
+    mlp_bda: Union[ModuleSpec, type] = IdentityFuncOp
+
+
+class FalconTransformerLayer(BaseLayer):
+    """A single transformer layer.
+
+    Transformer layer takes input with size [s, b, h] and returns an
+    output of the same size.
+        
+    """
+
+    def __init__(
+        self,
+        config: TransformerConfig,  # should come from FalconTransformerConfig class
+        submodules: FalconTransformerLayerSubmodules,
+        layer_number: int = 1,
+        self_attn_mask_type=AttnMaskType.padding,
+    ):
+        super().__init__(config=config, submodules=submodules)
+        self.config: TransformerConfig = config
+
+        self.layer_number = layer_number + self._get_layer_offset()
+
+        self.self_attn_mask_type = self_attn_mask_type
+
+        if hasattr(self.config, 'new_decoder_architecture'):
+            self.new_decoder_architecture = self.config.new_decoder_architecture
+        else:
+            self.new_decoder_architecture = None
+
+        if hasattr(self.config, 'parallel_attention'):
+            self.parallel_attention = self.config.parallel_attention
+        else:
+            self.parallel_attention = None
+
+        ## [Module 1: Input Layernorm] Optional Layernorm on the input data
+        # TODO: add pytorch only layernorm
+        self.input_layernorm = build_module(
+            submodules.input_layernorm,
+            config=self.config,
+            hidden_size=self.config.hidden_size,
+            eps=self.config.layernorm_epsilon,
+            persist_layer_norm=self.config.persist_layer_norm,
+            sequence_parallel=self.config.sequence_parallel,
+            zero_centered_gamma=self.config.layernorm_zero_centered_gamma,
+            normalization=self.config.normalization,
+        )
+
+        ## [Module 2: SelfAttention]
+        self.self_attention = build_module(submodules.self_attention, config=self.config, layer_number=layer_number,)
+
+        ## [Module 3: BiasDropoutFusion] Optional
```

Comment:
Is this optional? Was expecting Falcon to have a specific architecture.

Reply:
This part will be removed and directly use TransformerLayer's init